### PR TITLE
WIP: Improve error reporting: Recursive async functions 

### LINKED ISSUE #1170: Improve error reporting: Recursive async functions
## What
The following code snippet leads to the resulting error message: -
```fsharp
let rec foo() = async { return foo() }
```

```
Type mismatch. Expecting a
    'a    
but given a
    Async<'a>    
The resulting type would be infinite when unifying ''a' and 'Async<'a>'
```

## Why
99/100 times the cause is simply that the developer has forgotten to ```return!``` rather than ```return```. The current error message is confusing, does not explain why the error has occurred and does not present the likely solution.

## How
Suggest a rewording that follows *Error / Explanation / Suggestion* format: -

```
Type mismatch. Expecting a
    'a    
but given a
    Async<'a>
You are attempting to recursively return the result of an asynchronous call which is not awaited. Have you forgotten to use 'return!' instead of 'return' ?
```

Perhaps someone could explain what an infinite type is as well - I have no idea, although I sort of understand why this error occurs... :-) As I don't, the "explanation" bit of this error needs reworking for sure. I know that common parlance is to use "Consider" but in this case it's a guess more than anything. I also don't like use of "await" but am not sure what the correct term is.

### PR DESCRIPTION
**Pull request by forki**
- fixes #1170

This does two things:

* We don't talk about "infinite types" any more and just talk about unification issues.
* In the case of return we give advice to use return!

image

### CODE REVIEW

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -7635,6 +7635,11 @@ and TcComputationExpression cenv env overallTy mWhole interpExpr builderTy tpenv
         let mBuilderVal = mBuilderVal.MakeSynthetic()
         SynExpr.Lambda (false,false,SynSimplePats.SimplePats ([mkSynSimplePatVar false (mkSynId mBuilderVal builderValName)],mBuilderVal), runExpr, mBuilderVal)
 
+    let env =
+        match comp with     
+        | SynExpr.YieldOrReturn _ -> { env with eContextInfo = ContextInfo.ReturnInComputationExpression }
```
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: Will misuses of "yield" now report an error about "return"?  Test cases for that?

Comment (forki) [inline src/fsharp/TypeChecker.fs]: Can we distinguish the two?

**Comment by dsyme** [comment]
Awesome. The code looks just fine.  

I do still wonder if we should say something more than "types can't be unified", given we know there is a specific reason for this (one type would be included in the other, the types would be infinite etc.).

**Comment by forki** [comment]
yes we can improve on the wording. I just think we should avoid to talk about "infinite types" here. Even if this is the correct term it's probably a term that only very few people know. maybe we could use wording that talks about an "infinite loop while unifying". That is maybe a little less correct but I assume everyone has already seen an infinite loop.

File: src/fsharp/TypeChecker.fs (hunk 1 above)
Comment (forki) [inline src/fsharp/TypeChecker.fs]: uh just saw I didn't commit any of my tests. oups.

**Comment by dsyme** [comment]
@forki Any idea why AppVeyor is no longer showing in the list of CI builds?

**Comment by forki** [comment]
I think I saw a comment from @otawfik-ms that he removed the appveyor trigger.

File: src/fsharp/TypeChecker.fs (hunk 1 above)
Comment (forki) [inline src/fsharp/TypeChecker.fs]: @dsyme do you know a sample that would create infinite types using yield?!
And I think the code is only triggered from with CE that are not seq

**Comment by dsyme** [comment]
@forki Oh, ok.  

@otawfik-ms can you confirm? 

I don't mind as long as what we have is reliable - I'm just a creature of habit and I'm used to AppVeyor. I haven't learned Jenkins yet, but I'll start.  

@otawfik-ms Can you make sure either @forki or myself or both have admin access on Jenkins please?  And @KevinRansom too. Thanks.

**Comment by otawfik-ms** [comment]
@dsyme @forki  Jenkins is the main CI server for all dotnet projects now. Roslyn/other dotnet projects will be using it, with much better support for updates, parallel execution, and it is much faster (we are running all tests in CI now).
However, we will keep the appveyor script because a few community members suggested that they still want to run appveyor on their private forks for CI.

**Comment by forki** [comment]
just for fun, this is what elm gives for something similar:

image

[Source](https://twitter.com/GregorySchier/status/732830868562182144)

File: src/fsharp/TypeChecker.fs (hunk 1 above)
Comment (forki) [inline src/fsharp/TypeChecker.fs]: @Tamil gave the following code sample.
image

So indeed we have a false positive. Adding as test case and trying to fix it.

Comment (forki) [inline src/fsharp/TypeChecker.fs]: /// F# syntax: yield expr 
    /// F# syntax: return expr 
    /// Computation expressions only
    | YieldOrReturn   of (bool * bool) * SynExpr * range

@dsyme should we split that in ast.fs? Maybe in separate PR?

Comment (forki) [inline src/fsharp/TypeChecker.fs]: image

so that's good.

Comment (forki) [inline src/fsharp/TypeChecker.fs]: I leave 
image

for another day ;-)

**Comment by KevinRansom** [comment]
ci_part2 failed:

http://dotnet-ci.cloudapp.net/job/Microsoft_visualfsharp/job/master/job/release_ci_part2_windows_nt_prtest/348/consoleText

````
+++ Warnings (ReturnInsteadOfReturnBang.fs) +++
Microsoft (R) F# Compiler version (private)
Copyright (c) Microsoft Corporation. All Rights Reserved.
(4,32): error FS0001: Type mismatch. Expecting a
    ''a'    
but given a
    'Async<'a>'    
The types ''a' and 'Async<'a>' cannot be unified.

Consider using 'return!' instead of 'return'.

*** The following necessary lines were never matched:
***	\(4,32\):.+error FS0001:.+.+Consider using 'return!' instead of 'return'.*


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL
+++ Warnings (YieldInsteadOfYieldBang.fs) +++
Microsoft (R) F# Compiler version (private)
Copyright (c) Microsoft Corporation. All Rights Reserved.
(7,30): error FS0001: Type mismatch. Expecting a
    ''a'    
but given a
    ''a list'    
The types ''a' and ''a list' cannot be unified.

Consider using 'yield!' instead of 'yield'.

*** The following necessary lines were never matched:
***	\(7,30\):.+error FS0001:.+.+Consider using 'yield!' instead of 'yield'.*


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL







Error: 'RunTests.cmd release fsharpqa ' failed
````

**Comment by forki** [comment]
Yeah I'm trying like crazy to find the correct Regex. I think I tried like
15 times now

**Comment by forki** [comment]
@KevinRansom @enricosada any ideas why it doesn't match?

**Comment by forki** [comment]
this is seriously driving me nuts. I can't see the error in my regex.

**Comment by enricosada** [comment]
try multiple `<Expect` with single parts

```
//<Expects status="Error" span="(7,30)" id="FS0001">.Type mismatch. Expecting a.+''a'.+but given a.+''a list'.+The types ''a' and ''a list' cannot be unified. Consider using 'yield!' instead of 'yield'.*</Expects>
```

also span="(7,30)" is ok? i think the error message has more characters

**Comment by forki** [comment]
That's where it starts. Tbh I only want to check for the "consider" part.

**Comment by KevinRansom** [comment]
@forki I took care of the error messages:  
https://github.com/Microsoft/visualfsharp/commit/4cce423a8df094adadbc62becfa61470f152f78f

**Comment by forki** [comment]
Thanks so much Kevin. Didn't understand what was wrong.

**Pull request merged**